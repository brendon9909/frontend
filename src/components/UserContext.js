import { createContext, useState } from "react";

const UserContext = createContext();

export const UserProvider = ({ children }) => {
  const [token, setToken] = useState(null);
  const [username, setUsername] = useState(null)
  const [task, setTask] = useState(null)

  return (
    <UserContext.Provider value={{ token, setToken, username, setUsername, task, setTask }}>
      {children}
    </UserContext.Provider>
  );
};

export default UserContext;
