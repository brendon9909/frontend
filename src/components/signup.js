import { Button } from "react-bootstrap";
import { useFormik } from "formik";
import { NavLink, useNavigate } from 'react-router-dom';
import { useState } from "react";

export default function Signup() {
  const [myError, setMyError] = useState(null)
  const navigate = useNavigate()
  const formik = useFormik({
    initialValues: {
      username: "",
      password: "",
    },
    onSubmit: async (values) => {
      try {
        const response = await fetch("http://localhost:8000/signup", {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify(values),
        });

        if (response.ok) {
          setMyError(null)
          navigate('/login')
        }else{
          const responseData = await response.json()
          setMyError(responseData.message)
        }
      } catch (error) {
        console.error(error);
      }
    },
    validate: (values) => {
      const errors = {}

      if(!values.username){
        errors.username = "Required"
      }

      if(!values.password){
        errors.password = "Required"
      }

      return errors;
    }
  });

  return (
      <div className="signup">
        <h1>Signup</h1>
        <hr />
        <form onSubmit={formik.handleSubmit}>
          <label className="labels" htmlFor="username">
            Username
          </label>
          <br />
          <input
            className="inpBox"
            name="username"
            type="text"
            placeholder="username"
            value={formik.values.username}
            onChange={formik.handleChange}
          />{myError ? <div className="error">{myError}</div>: null}
          {formik.touched.username && formik.errors.username ? <div className="error">{formik.errors.username}</div>: null}
          <br />
          <br />
          <label className="labels" htmlFor="password">
            Password
          </label>
          <br />
          <input
            className="inpBox"
            name="password"
            type="password"
            placeholder="*******"
            value={formik.values.password}
            onChange={formik.handleChange}
          />{formik.touched.password && formik.errors.password ? <div className="error">{formik.errors.password}</div>: null}
          <br />
          <br />
          <Button type="submit" style={{ width: "20vh" }} variant="success">
            Signup
          </Button>
        </form>
        <p>Already have an account? Login <NavLink to="/login">here</NavLink></p>
      </div>
  );
}
