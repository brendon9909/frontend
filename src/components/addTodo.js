import { Button, Modal } from "react-bootstrap";
import { useFormik } from "formik";
import UserContext from "./UserContext";
import { useContext, useState } from "react";

//function to add new todos
export default function AddTodo(props) {
  //get the stored username
  const { username } = useContext(UserContext);
  //create error state to store error from backend
  const [myError, setMyError] = useState(null);
  //formik to handle the add todo form
  const formik = useFormik({
    initialValues: {
      task: "",
    },
    onSubmit: async (values) => {
      try {
        //make the put request to api
        const response = await fetch("http://localhost:8000/todo/add", {
          method: "PUT",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify({
            username: username,
            todo: values.task,
          }),
        });

        if (response.ok) {
          //close the modal
          props.close();
          //reload the todos to reflect new todo
          props.reload();
          //clear the error
          setMyError(null);
        } else {
          //create the error
          const responseData = await response.json();
          setMyError(responseData.message);
        }
      } catch (error) {
        console.error("something went wrong", error);
      }
    },
    validate: (values) => {
      //formik validating
      const errors = {};

      if (!values.task) {
        errors.task = "required";
      }

      return errors;
    },
  });

  return (
    <Modal show={props.show}>
      {/*modal for adding new todos*/}
      <Modal.Header>Add New Todo</Modal.Header>
      <Modal.Body>
        <form onSubmit={formik.handleSubmit}>
          <label htmlFor="task">Task</label>
          <br />
          <input
            type="text"
            placeholder="task"
            name="task"
            value={formik.values.task}
            onChange={formik.handleChange}
          />
          {formik.touched.task && formik.errors.task ? (
            <div className="error">{formik.errors.task}</div>
          ) : null}
          {myError ? <div className="error">{myError}</div> : null}
          <br />
          <Button type="submit" variant="primary">
            Add Task
          </Button>
        </form>
      </Modal.Body>
      <Modal.Footer>
        <Button variant="danger" onClick={props.close}>
          Cancel
        </Button>
      </Modal.Footer>
    </Modal>
  );
}
