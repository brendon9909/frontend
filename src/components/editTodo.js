import { useFormik } from "formik";
import { useContext, useState } from "react";
import { Button, Modal } from "react-bootstrap";
import UserContext from "./UserContext";

//function for editing a todo
export default function EditTodo(props) {
  //get the task stored globally
  const { task, username } = useContext(UserContext);
  //create error state
  const [myError, setMyError] = useState(null);

  const formik = useFormik({
    //formik for handling editing tasks
    initialValues: {
      task: "",
    },
    onSubmit: async (values) => {
      try {
        //make the put request
        const response = await fetch("http://localhost:8000/todo/edit", {
          method: "PUT",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify({
            username: username,
            todo: task,
            newTodo: values.task,
          }),
        });

        if (response.ok) {
          //reload todos
          props.reset();
          //close the modal
          props.close();
          //clear the error
          setMyError(null);
        } else {
          //create the error
          const responseData = await response.json();
          setMyError(responseData.message);
        }
      } catch (error) {
        console.error("something went wrong", error);
      }
    },
    validate: (values) => {
      const errors = {};

      if (!values.task) {
        errors.task = "Required";
      }

      return errors;
    },
  });

  return (
    <Modal show={props.show}>
      {/*modal for editing a task*/}
      <Modal.Header>Edit Task</Modal.Header>
      <Modal.Body>
        <form onSubmit={formik.handleSubmit}>
          <label htmlFor="task">Edit</label>
          <br />
          <input
            name="task"
            type="text"
            value={formik.values.task}
            onChange={formik.handleChange}
          />
          {formik.touched.task && formik.errors.task ? (
            <div className="error">{formik.errors.task}</div>
          ) : null}
          {myError ? <div className="error">{myError}</div> : null}
          <br />
          <Button variant="success" type="submit">
            Apply changes
          </Button>
        </form>
      </Modal.Body>
      <Modal.Footer>
        <Button variant="danger" onClick={props.close}>
          Close
        </Button>
      </Modal.Footer>
    </Modal>
  );
}
