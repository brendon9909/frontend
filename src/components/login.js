import { useFormik } from "formik";
import { Button } from "react-bootstrap";
import { NavLink, useNavigate } from "react-router-dom";
import UserContext from "./UserContext";
import { useContext, useState } from "react";

//login page
export default function Login() {
  //context for setting the token and username globally
  const { setToken, setUsername } = useContext(UserContext);
  //state for storing error from backend
  const [myError, setMyError] = useState(null)

  const navigate = useNavigate();

  const formik = useFormik({
    //formik for form handling the login
    initialValues: {
      username: "",
      password: "",
    },
    onSubmit: async (values) => {
      try {
        //make the login request
        const response = await fetch("http://localhost:8000/login", {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify(values),
        });

        if (response.ok) {
          const responseData = await response.json();
          //set the username
          setUsername(responseData.user);
          //set the token
          setToken(responseData.token);
          //go to home page
          navigate("/home");
          //clear error
          setMyError(null)
        }
        else{
          const responseData = await response.json()
          //set the error
          setMyError(responseData.message)
        }
      } catch (error) {
        console.error(error.message);
      }
    },
    validate: (values) => {
      //formik error validating
      const errors = {};

      if (!values.username) {
        errors.username = "Required";
      }

      if (!values.password) {
        errors.password = "Required";
      }

      return errors;
    },
  });

  return (
    <div className="signup">
      <h1>Login</h1>
      <hr />
      <form onSubmit={formik.handleSubmit}>
        <label className="labels" htmlFor="username">
          Username
        </label>
        <br />
        <input
          className="inpBox"
          name="username"
          type="text"
          placeholder="username"
          value={formik.values.username}
          onChange={formik.handleChange}
        />
        {formik.touched.username && formik.errors.username ? (
          <div className="error">{formik.errors.username}</div>
        ) : null}
        {myError ? <div className="error">{myError}</div>: null}
        <br />
        <br />
        <label className="labels" htmlFor="password">
          Password
        </label>
        <br />
        <input
          className="inpBox"
          name="password"
          type="password"
          placeholder="*******"
          value={formik.values.password}
          onChange={formik.handleChange}
        />
        {formik.touched.password && formik.errors.password ? (
          <div className="error">{formik.errors.password}</div>
        ) : null}
        <br />
        <br />
        <Button type="submit" style={{ width: "20vh" }} variant="success">
          Login
        </Button>
      </form>
      <p>
        Don't have an account? Signup <NavLink to="/">here</NavLink>
      </p>
    </div>
  );
}
