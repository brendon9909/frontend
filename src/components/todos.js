import { useContext, useEffect, useState } from "react";
import UserContext from "./UserContext";
import { useNavigate } from "react-router-dom";
import { Button } from "react-bootstrap";
import AddTodo from "./addTodo";
import EditTodo from "./editTodo";

export default function Todos() {
  const navigate = useNavigate();
  //get the token, username and setTask form global storage
  const { token, username, setTask } = useContext(UserContext);
  //create todos state to store all todos
  const [todos, setTodos] = useState([]);
  //state to show add todo modal
  const [show, setShow] = useState(false);
  //state to show edit todo modal
  const [showEdit, setShowEdit] = useState(false);

  //close the add todo and edit todo modals
  const handleClose = () => setShow(false);
  const handleEditClose = () => setShowEdit(false);

  useEffect(() => {
    //check if there is a token and load todos if there is
    if (!token) {
      navigate("/login");
    } else {
      fetchData();
    }
  }, []);

  async function fetchData() {
    //function to load all todos using the token
    try {
      //get request to api
      const response = await fetch("http://localhost:8000/login/data", {
        method: "GET",
        headers: {
          Authorization: `Bearer ${token}`,
        },
      });

      if (response.ok) {
        //set todos state as our response data
        const responseData = await response.json();
        setTodos(responseData);
      }
    } catch (error) {
      console.error(error.message);
    }
  }

  const markComplete = async (todo) => {
    //make a request to set a specific todo as complete
    try {
      const response = await fetch("http://localhost:8000/todo/complete", {
        method: "PUT",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          username: username,
          todo: todo,
        }),
      });

      if (response.ok) {
        //reload the todos
        fetchData();
      }
    } catch (error) {
      console.error("something went wrong", error);
    }
  };

  const markIncomplete = async (todo) => {
    //request to make a spcific todo incomplete
    try {
      const response = await fetch("http://localhost:8000/todo/reset", {
        method: "PUT",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          username: username,
          todo: todo,
        }),
      });

      if (response.ok) {
        //reload the todos
        fetchData();
      }
    } catch (error) {
      console.error("something went wrong", error);
    }
  };

  const removeTodo = async (todo) => {
    //function to delete a todo
    try {
      //delete request
      const response = await fetch("http://localhost:8000/todo/delete", {
        method: "DELETE",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          username: username,
          todo: todo,
        }),
      });

      if (response) {
        //reload todos
        fetchData();
      }
    } catch (error) {
      console.error("something went wrong", error);
    }
  };

  return (
    <div style={{ backgroundColor: "black", padding: "51px", color: "white" }}>
      <h1 style={{ fontFamily: "Protest Guerrilla" }}>Todos</h1>
      <div className="todos">
        {todos.length > 0 ? (
          <>
            {/*map the todos*/}
            {todos.map((item) => (
              <div style={{ display: "flex" }}>
                <h5
                  style={{
                    textDecoration: item.completed ? "line-through" : "none",
                  }}
                >
                  <input
                    type="checkbox"
                    id={item.text}
                    onClick={() => {
                      let checkbox = document.getElementById(item.text);

                      if (checkbox.checked) {
                        markComplete(item.text);
                      } else {
                        markIncomplete(item.text);
                      }
                    }}
                    checked={item.completed}
                  />{" "}
                  {item.text}
                  <img
                    className="deleteIcon"
                    onClick={() => removeTodo(item.text)}
                    src="https://cdn2.iconfinder.com/data/icons/thin-line-color-1/21/33-512.png"
                    alt="delete"
                  />
                  <img
                    className="editIcon"
                    onClick={() => {
                      setTask(item.text);
                      setShowEdit(true);
                    }}
                    src="https://creazilla-store.fra1.digitaloceanspaces.com/icons/3271683/edit-icon-md.png"
                    alt="edit"
                  />
                </h5>
              </div>
            ))}
          </>
        ) : (
          <h4>Start adding todos</h4>
        )}
        <br />
        <Button
          variant="outline-success"
          onClick={() => {
            setShow(true);
          }}
        >
          Add new Todo
        </Button>
      </div>
      <AddTodo show={show} close={handleClose} reload={fetchData} />
      <EditTodo show={showEdit} close={handleEditClose} reset={fetchData} />
    </div>
  );
}
