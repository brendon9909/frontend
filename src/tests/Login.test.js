import Login from "../components/login";
import React from "react";
import Renderer from "react-test-renderer";
import { MemoryRouter } from "react-router-dom";
import { UserProvider } from "../components/UserContext";

test("Login renders correctly", () => {
  const tree = Renderer.create(
    <UserProvider>
      <MemoryRouter>
        <Login />
      </MemoryRouter>
    </UserProvider>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
