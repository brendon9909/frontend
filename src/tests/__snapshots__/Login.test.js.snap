// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders correctly 1`] = `
<div
  className="signup"
>
  <h1>
    Login
  </h1>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <label
      className="labels"
      htmlFor="username"
    >
      Username
    </label>
    <br />
    <input
      className="inpBox"
      name="username"
      onChange={[Function]}
      placeholder="username"
      type="text"
      value=""
    />
    <br />
    <br />
    <label
      className="labels"
      htmlFor="password"
    >
      Password
    </label>
    <br />
    <input
      className="inpBox"
      name="password"
      onChange={[Function]}
      placeholder="*******"
      type="password"
      value=""
    />
    <br />
    <br />
    <button
      className="btn btn-success"
      disabled={false}
      style={
        Object {
          "width": "20vh",
        }
      }
      type="submit"
    >
      Login
    </button>
  </form>
  <p>
    Don't have an account? Signup 
    <a
      aria-current="page"
      className="active"
      href="/"
      onClick={[Function]}
    >
      here
    </a>
  </p>
</div>
`;
